src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatabilityAnalyzer.cs

```diff
@@ -0,0 +1,446 @@
+            }
+        }
+
+        private static bool IsWithinConditionalOperation(IOperation pOperation) =>
+            pOperation.ConstantValue.HasValue &&
+            pOperation.Parent is IBinaryOperation bo &&
+            (bo.OperatorKind == BinaryOperatorKind.Equals ||
+            bo.OperatorKind == BinaryOperatorKind.NotEquals ||
+            bo.OperatorKind == BinaryOperatorKind.GreaterThan ||
+            bo.OperatorKind == BinaryOperatorKind.LessThan ||
+            bo.OperatorKind == BinaryOperatorKind.GreaterThanOrEqual ||
+            bo.OperatorKind == BinaryOperatorKind.LessThanOrEqual);
+
+        private static bool IsOSPlatformsEqual(string left, string right)
+        {
+            if (left.Equals(right, StringComparison.OrdinalIgnoreCase))
+            {
+                return true;
+            }
+            else
+            {
+                var normalizedLeft = NormalizeOSName(left);
+                var normalizedRight = NormalizeOSName(right);
+                return string.Equals(normalizedLeft, normalizedRight, StringComparison.OrdinalIgnoreCase);
+            }
+        }
+
+        private static string NormalizeOSName(string name)
+        {
+            if (name.Equals(OSX, StringComparison.OrdinalIgnoreCase))
+                return macOS;
+            return name;
+        }
+
+        private static DiagnosticDescriptor SelectRule(PlatformAttributeType attributeType)
+        {
+            if (attributeType == PlatformAttributeType.MinimumOSPlatformAttribute)
+                return MinimumOsRule;
+            if (attributeType == PlatformAttributeType.ObsoletedInOSPlatformAttribute)
+                return ObsoleteRule;
+            return RemovedRule;
+        }
+
+        private static AttributeData[] FindAllPlatformAttributesApplied(ImmutableArray<AttributeData> immediateAttributes, INamedTypeSymbol parent)
+        {
+            using var builder = ArrayBuilder<AttributeData>.GetInstance();
+            foreach (AttributeData attribute in immediateAttributes)
+            {
+                if (s_osPlatformAttributes.Contains(attribute.AttributeClass.Name))
+                {
+                    builder.Add(attribute);
+                }
+            }
+
+            while (parent != null)
+            {
+                var current = parent.GetAttributes();
+                foreach (var attribute in current)
+                {
+                    if (s_osPlatformAttributes.Contains(attribute.AttributeClass.Name) &&
+                        TargetPlatformAttribute != attribute.AttributeClass.Name)
+                    {
+                        builder.Add(attribute);
+                    }
+                }
+                parent = parent.BaseType;
+            }
+            return builder.ToArray();
+        }
+
+        private static bool IsSuppressedByAttribute(PlatformAttributeInfo diagnosingAttribute, ISymbol containingSymbol)
```

Comment:
As you had initially predicted, this walk would be pretty expensive without caching applicable platform information for symbols. It is fine to not do it as part of this PR, but absolutely need it before shipping. Please also thoroughly measure build performance, with and without this analyzer to confirm the impact.

Reply:
Yeah now its time to look at them